src/mlir/lowering.md

```diff
@@ -0,0 +1,517 @@
+form of AST nodes, IR or dialect operations.
+
+If the calls depend on the type, for example which type a literal string will be
+stored as (ascii, UTF-8, etc) will change which functionality to call from the
+libraries and even the method names, which may not be known at a lower level.
+
+A clean implementation, and in the spirit of MLIR, is to have implementation
+details as dialect constructs (ops, types, attributes) that can later be easily
+lowered to a call to a library function using the available information.
+
+If the code is elided, no information is kept, no harm done. If the types change,
+the information is carried with the type, and the lowering is still trivial.
+
+# Structure
+
+## Overall structure
+
+The current 1IR design has the AST as the main structure, with the basic blocks
+in IR form. This should change to a pure IR representation in time, but even
+before that, we can use the AST class structure to lower structs for fields and
+mangle names for functions only, with the rest lowered directly from 1IR as the
+body of methods.
+
+Class and module lexical context will be controlled with MLIR's `module`
+construct, with their regions containing a list of methods that belong to that
+class/module.
+
+Functions will be lowered as MLIR functions, called with `std.call` for static
+functions that are resolvable via MLIR symbol table and `verona.call` for the
+dynamic calls that will later be converted into vtable lookups (made concrete
+only after colouring).
+
+Lambdas and `when` blocks will have already been converted into an anonymous
+class with captures as fields and the `apply` method as the body. Constant
+literals will also be lowered in that way, with the initialisation as the
+`construct` method and the value as an immutable field.
+
+Due to reachability analysis being done in the 1IR, by the time we get it to
+lower to 2IR we won't have any generics. All methods and fields will be
+guaranteed to be used/called in the program from the entry point and all the
+types will be concrete and correct.
+
+## Function Bodies
+
+Function bodies have multiple `BodyIR` regions, one for the main body and one
+for each `when` clauses. The main body will be lowered in the context of the
+encompassing class (struct field access, mangling) while the `when` bodies will
+by lowered on their new anonymous classes and captures, still within the same
+context, so to guarantee no external user can access it by mistake.
+
+## Basic Blocks
+
+So far, there is only two constructs that create new basic blocks: conditionals
+and loops. Both will end with a tail block that will be the main block of the
+subsequent operations.
+
+The main issue when lowering basic blocks is figuring out their arguments. For
+this we need free variable and liveness analysis, which will have already been
+done by the 1IR passes. That information is in the `FnAnalysis` structure for
+now and it should be trivial to use it for arguments without having to backtrack.
```

Comment:
The 1IR is already in SSA form and has all the basic block arguments you need. There's no need to worry about that anymore.